### PR TITLE
support skip-index based on chunk-metrics to accelerate expr filter(#27925)

### PR DESCRIPTION
**Pull request by MrPresent-Han**
related: #27925

### CODE REVIEW

File: internal/core/src/index/ScalarIndexSort-inl.h [hunk 1]
```diff
@@ -307,4 +313,44 @@ ScalarIndexSort<T>::Reverse_Lookup(size_t idx) const {
     auto offset = idx_to_offsets_[idx];
     return data_[offset].a_;
 }
+
```
Comment (MrPresent-Han) [inline internal/core/src/index/ScalarIndexSort-inl.h:316]: aggregate unary and binary judgment into one function

File: internal/core/src/query/visitors/ExecExprVisitor.cpp [hunk 2]
```diff
@@ -293,11 +302,18 @@ ExecExprVisitor::ExecRangeVisitorImpl(FieldId field_id,
                    "[ExecExprVisitor]Data size not equal to size_per_chunk");
         results.emplace_back(std::move(data));
     }
+
     for (auto chunk_id = indexing_barrier; chunk_id < num_chunk; ++chunk_id) {
         auto this_size = chunk_id == num_chunk - 1
                              ? row_count_ - chunk_id * size_per_chunk
                              : size_per_chunk;
```
Comment (MrPresent-Han) [inline internal/core/src/query/visitors/ExecExprVisitor.cpp:309]: use skip Index here when executing expr

File: internal/core/src/segcore/SegmentInterface.cpp [hunk 3]
```diff
@@ -293,4 +293,19 @@ SegmentInternalInterface::timestamp_filter(BitsetType& bitset,
     }
 }
```
Comment (MrPresent-Han) [inline internal/core/src/segcore/SegmentInterface.cpp:294]: at present, SkipIndex is a segment-level component, and relating load is also triggered by segment

File: internal/core/src/segcore/SegmentInterface.cpp [hunk 4]
```diff
@@ -293,4 +293,19 @@ SegmentInternalInterface::timestamp_filter(BitsetType& bitset,
...
+void
+SegmentInternalInterface::LoadSkipIndex(milvus::FieldId field_id,
+                                        int64_t chunk_id,
+                                        milvus::DataType data_type,
+                                        const void *chunk_data,
+                                        int64_t count) {
+    skipIndex_.Load(field_id, chunk_id, data_type, chunk_data, count);
+}
+
+
```
Comment (MrPresent-Han) [inline internal/core/src/segcore/SegmentInterface.cpp:310]: maybe we need a returned value to indicate whether successful or not for loading skipIndex

File: internal/core/src/segcore/SkipIndex.h [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+};
+
+struct FieldChunkMetrics {
+    MinValue min;
+    MaxValue max;
+    bool hasValue_;
+
+    FieldChunkMetrics() : hasValue_(false){};
+};
+
```
Comment (MrPresent-Han) [inline internal/core/src/segcore/SkipIndex.h:45]: SkipIndex is a separate class and can be extended for using fieldChunkMetrics in various ways.

File: internal/core/src/segcore/SkipIndex.h [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+        static constexpr bool isNumericType =
+                std::is_same<T, int8_t>::value || std::is_same<T, int16_t>::value ||
+                std::is_same<T, int32_t>::value ||
+                std::is_same<T, int64_t>::value || std::is_same<T, float>::value ||
+                std::is_same<T, double>::value;
+        static constexpr bool isDisabledType =
+                std::is_same<T, std::string>::value || std::is_same<T, milvus::Json>::value;
+        static constexpr bool value = isNumericType && !isDisabledType;
+    };
+
```
Comment (MrPresent-Han) [inline internal/core/src/segcore/SkipIndex.h:97]: one C++ question here, how can I avoid pass DataType by using T simply?

File: internal/core/unittest/test_sealed.cpp [hunk 7]
```diff
@@ -1381,3 +1381,98 @@ TEST(Sealed, LoadArrayFieldDataWithMMap) {
     SealedLoadFieldData(dataset, *segment, {}, true);
     segment->Search(plan.get(), ph_group.get());
 }
+
```
Comment (MrPresent-Han) [inline internal/core/unittest/test_sealed.cpp:1384]: old ut, I ' m changing now

File: internal/core/src/segcore/SkipIndex.h [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+    Load(milvus::FieldId field_id,
+         int64_t chunk_id,
+         milvus::DataType data_type,
+         const void *chunk_data,
+         int64_t count);
+
+private:
+    const FieldChunkMetrics&
+    GetFieldChunkMetrics(FieldId field_id, int chunk_id) const;
+
```
Comment (MrPresent-Han) [inline internal/core/src/segcore/SkipIndex.h:85]: one C++ question, it's weird that when using std::is_integral<T> and std::floating_point to replace all these std::is_same<T, ***>, the function cannot be instantiated correctly and complition failed.

**Review by MrPresent-Han — COMMENTED** [review]
need review for interface design and basic functions has been testified correctly on my local macOS

**Comment by MrPresent-Han** [comment]
rerun ut

File: internal/core/src/query/visitors/ExecExprVisitor.cpp [hunk 9]
```diff
@@ -1641,8 +1749,11 @@ ExecExprVisitor::visit(UnaryRangeExpr& expr) {
         }
         case DataType::VARCHAR: {
             if (segment_.type() == SegmentType::Growing) {
+                std::cout << "hc===execute unary range: string" << std::endl;
```
Comment (jiaoew1991) [inline internal/core/src/query/visitors/ExecExprVisitor.cpp:1752]: cleanup it

**Comment by MrPresent-Han** [comment]
rerun ut

File: internal/core/src/query/visitors/ExecExprVisitor.cpp (hunk 9 above)
Comment (MrPresent-Han) [reply internal/core/src/query/visitors/ExecExprVisitor.cpp]: changed

**Review by yah01 — COMMENTED** [review]
skip for term expr would come later?

**Comment by MrPresent-Han** [comment]
> skip for term expr would come later?

emm， I forget this, good catch

**Comment by MrPresent-Han** [comment]
> skip for term expr would come later?

changed

File: internal/core/src/query/visitors/ExecExprVisitor.cpp [hunk 10]
```diff
@@ -2522,9 +2633,19 @@ ExecExprVisitor::ExecTermVisitorImplTemplate<bool>(TermExpr& expr_raw)
         // return std::binary_search(terms.begin(), terms.end(), x);
         return term_set.find(x) != term_set.end();
     };
+    auto skip_index_func =
```
Comment (MrPresent-Han) [inline internal/core/src/query/visitors/ExecExprVisitor.cpp:2636]: changed, support term @yah01

**Review by MrPresent-Han — COMMENTED** [review]
support skip term, need review

**Review by yah01 — COMMENTED** [review]
/lgtm

**Pull request merged**